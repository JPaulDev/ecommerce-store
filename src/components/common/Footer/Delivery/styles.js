import styled from 'styled-components';

export const Container = styled.div`
  @media screen and (min-width: 650px) {
    margin-bottom: 30px;
    margin-right: 30px;
  }

  @media screen and (min-width: 860px) {
    margin-right: 50px;
  }

  @media screen and (min-width: 1050px) {
    margin: 0;
  }
`;

export const Text = styled.p`
  color: var(--color-grey-8);
  font-size: 0.82rem;
  margin: 8px 0 8px 20px;

  @media screen and (min-width: 650px) {
    margin: 8px 0;
  }
`;

export const ButtonContainer = styled.div`
  display: flex;
`;

export const TabButton = styled.button`
  background-color: ${({ active }) =>
    active ? 'var(--color-grey-11)' : '#dedede'};
  border-bottom: ${({ active }) =>
    active ? '3px solid var(--color-orange-1)' : '3px solid #dedede'};
  cursor: pointer;
  flex: 1;
  font-weight: 600;
  padding: 8px;
  transition: all 200ms ease-in-out;
`;

export const Form = styled.form`
  align-items: center;
  background-color: var(--color-grey-11);
  border-bottom: 1px solid var(--color-grey-10);
  display: flex;
  padding: 20px 15px;

  @media screen and (max-width: 649px) {
    justify-content: center;
  }
`;

export const Input = styled.input`
  border: 1px solid grey;
  border-radius: 5px;
  margin-right: 8px;
  max-width: 175px;
  padding: 8px 0 8px 10px;
  width: 40%;
`;

export const SubmitButton = styled.button`
  align-items: center;
  background: var(--color-grey-1);
  border-radius: 50%;
  cursor: pointer;
  display: flex;
  height: 32px;
  justify-content: center;
  width: 32px;

  &:active {
    transform: scale(0.9);
  }
`;
