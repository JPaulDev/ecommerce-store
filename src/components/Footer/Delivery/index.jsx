import { useState, useEffect, useId } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import useMediaQuery from '../../../hooks/useMediaQuery';
import MaterialIcon from '../../MaterialIcon';
import { Heading, DropdownButton } from '../styles';
import * as Styled from './styles';

const forms = [
  {
    action: 'https://pc-connect.co.uk/tracking/trackorder',
    placeholder: 'Postcode',
    name: 'postcode',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackreturn',
    placeholder: 'RMA No.',
    name: 'rmaNumber',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackbuild',
    placeholder: 'Build No.',
    name: 'buildNumber',
  },
];

export default function Delivery() {
  const [activeTab, setActiveTab] = useState(0);
  const [isOpen, setIsOpen] = useState(window.innerWidth >= 650);
  const isMatch = useMediaQuery('(min-width: 650px)');
  const id = useId();
  const heading = 'Delivery / Tracking Information';

  const handleSetActiveTab = (index) => setActiveTab(index);
  const handleOpen = () => setIsOpen(!isOpen);

  // Prevents the menu being stuck closed when resizing from a smaller window width
  useEffect(() => {
    isMatch ? setIsOpen(true) : setIsOpen(false);
  }, [isMatch]);

  return (
    <Styled.Container>
      {isMatch ? (
        <Heading>{heading}</Heading>
      ) : (
        <DropdownButton
          type="button"
          isOpen={isOpen}
          aria-expanded={isOpen}
          aria-controls="delivery"
          onClick={handleOpen}
        >
          {heading}
          <MaterialIcon iconSize="30px">chevron_right</MaterialIcon>
        </DropdownButton>
      )}
      <AnimatePresence initial={false}>
        {isOpen && (
          <motion.div
            id="delivery"
            initial={{ height: 0 }}
            animate={{ height: 'auto' }}
            exit={{ height: 0 }}
            transition={{ ease: 'easeInOut', duration: 0.5 }}
          >
            <Styled.Text>Order now to receive by tomorrow.</Styled.Text>
            <Styled.ButtonContainer>
              {['Order', 'Return', 'System'].map((text, index) => (
                <Styled.TabButton
                  key={text}
                  active={index === activeTab}
                  onClick={() => handleSetActiveTab(index)}
                  type="button"
                >
                  {text}
                </Styled.TabButton>
              ))}
            </Styled.ButtonContainer>
            {forms.map((item, index) => {
              if (index === activeTab) {
                return (
                  <Styled.Form
                    key={item.name}
                    action={item.action}
                    method="post"
                  >
                    <label htmlFor={`${id}-invoice`} hidden>
                      Invoice Number
                    </label>
                    <Styled.Input
                      id={`${id}-invoice`}
                      name="invoiceNumber"
                      autoComplete="off"
                      type="text"
                      maxLength="10"
                      placeholder="Invoice No."
                      required
                    />
                    <label htmlFor={id} hidden>
                      {item.placeholder.replace('No.', 'Number')}
                    </label>
                    <Styled.Input
                      id={id}
                      name={item.name}
                      autoComplete="off"
                      type="text"
                      maxLength="10"
                      placeholder={item.placeholder}
                      required
                    />
                    <Styled.SubmitButton
                      type="submit"
                      title="Submit"
                      aria-label="Submit"
                    >
                      <MaterialIcon iconSize="32px" iconColor="white">
                        chevron_right
                      </MaterialIcon>
                    </Styled.SubmitButton>
                  </Styled.Form>
                );
              }
              return undefined;
            })}
          </motion.div>
        )}
      </AnimatePresence>
    </Styled.Container>
  );
}
