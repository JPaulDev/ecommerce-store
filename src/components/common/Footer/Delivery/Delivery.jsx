import { useState, useEffect, useId } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import useMediaQuery from '../../../../hooks/useMediaQuery';
import ChevronRight from '../../../icons/ChevronRight';
import { Heading, DropdownButton } from '../styles';
import * as Styled from './styles';

const forms = [
  {
    action: 'https://pc-connect.co.uk/tracking/trackorder',
    placeholder: 'Postcode',
    name: 'postcode',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackreturn',
    placeholder: 'RMA No.',
    name: 'rmaNumber',
  },
  {
    action: 'https://pc-connect.co.uk/tracking/trackbuild',
    placeholder: 'Build No.',
    name: 'buildNumber',
  },
];

const variants = {
  open: {
    overflow: 'hidden',
    height: 'auto',
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
    },
  },
  closed: {
    overflow: 'hidden',
    height: 0,
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
    },
  },
};

export default function Delivery() {
  const [activeTab, setActiveTab] = useState(0);
  const [isOpen, setIsOpen] = useState(false);
  const isMatch = useMediaQuery('(min-width: 650px)');
  const id = useId();
  const heading = 'Delivery / Tracking Information';

  const handleSetActiveTab = (index) => setActiveTab(index);
  const handleOpen = () => setIsOpen(!isOpen);

  // Prevents the menu being stuck closed when resizing from a smaller window width
  useEffect(() => {
    isMatch ? setIsOpen(true) : setIsOpen(false);
  }, [isMatch]);

  return (
    <Styled.Container>
      {isMatch ? (
        <Heading>{heading}</Heading>
      ) : (
        <DropdownButton
          type="button"
          isOpen={isOpen}
          aria-expanded={isOpen}
          aria-controls="delivery"
          onClick={handleOpen}
        >
          {heading}
          <ChevronRight width={15} height={20} />
        </DropdownButton>
      )}
      <AnimatePresence>
        {isOpen && (
          <motion.div
            id="delivery"
            variants={variants}
            initial="closed"
            animate="open"
            exit="closed"
          >
            <Styled.Text>Order now to receive by tomorrow.</Styled.Text>
            <Styled.ButtonContainer>
              {['Order', 'Return', 'System'].map((text, index) => (
                <Styled.TabButton
                  key={text}
                  active={index === activeTab}
                  onClick={() => handleSetActiveTab(index)}
                  type="button"
                >
                  {text}
                </Styled.TabButton>
              ))}
            </Styled.ButtonContainer>
            {forms.map((item, index) => {
              if (index === activeTab) {
                return (
                  <Styled.Form
                    key={item.name}
                    action={item.action}
                    method="post"
                  >
                    <label htmlFor={`${id}-invoice`} hidden>
                      Invoice Number
                    </label>
                    <Styled.Input
                      id={`${id}-invoice`}
                      name="invoiceNumber"
                      autoComplete="off"
                      type="text"
                      maxLength="10"
                      placeholder="Invoice No."
                      required
                    />
                    <label htmlFor={id} hidden>
                      {item.placeholder.replace('No.', 'Number')}
                    </label>
                    <Styled.Input
                      id={id}
                      name={item.name}
                      autoComplete="off"
                      type="text"
                      maxLength="10"
                      placeholder={item.placeholder}
                      required
                    />
                    <Styled.SubmitButton
                      type="submit"
                      title="Submit"
                      aria-label="Submit"
                    >
                      <ChevronRight width={14} height={19} fill="white" />
                    </Styled.SubmitButton>
                  </Styled.Form>
                );
              }
              return undefined;
            })}
          </motion.div>
        )}
      </AnimatePresence>
    </Styled.Container>
  );
}
