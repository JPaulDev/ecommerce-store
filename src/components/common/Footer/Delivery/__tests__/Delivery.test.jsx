import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Delivery from '../index';

describe('Delivery component', () => {
  it('renders an order tab where order tracking can be checked', () => {
    const { asFragment } = render(<Delivery />);
    const button = screen.getByRole('button', {
      name: /order/i,
    });

    userEvent.click(button);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a return tab where returns status can be checked', () => {
    const { asFragment } = render(<Delivery />);
    const button = screen.getByRole('button', {
      name: /return/i,
    });

    userEvent.click(button);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a system build tab where build progress can be checked', () => {
    const { asFragment } = render(<Delivery />);
    const button = screen.getByRole('button', {
      name: /system/i,
    });

    userEvent.click(button);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Delivery component text inputs', () => {
  it('correctly records user input', () => {
    render(<Delivery />);
    const invoice = screen.getByLabelText(/invoice number/i);
    const postcode = screen.getByLabelText(/postcode/i);

    expect(invoice).toHaveValue('');
    expect(postcode).toHaveValue('');

    userEvent.type(invoice, '12345');
    userEvent.type(postcode, 'abcde');

    expect(invoice).toHaveValue('12345');
    expect(postcode).toHaveValue('abcde');
  });
});
