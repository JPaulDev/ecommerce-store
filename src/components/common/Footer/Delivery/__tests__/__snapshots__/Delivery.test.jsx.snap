// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delivery component renders a return tab where returns status can be checked 1`] = `
<DocumentFragment>
  <div
    class="sc-44f25114-0 egPNsa"
  >
    <h3
      class="sc-9bf04a45-3 jXmISH"
    >
      Delivery / Tracking Information
    </h3>
    <div
      id="delivery"
      style="overflow: hidden; height: auto;"
    >
      <p
        class="sc-44f25114-1 cwwZoe"
      >
        Order now to receive by tomorrow.
      </p>
      <div
        class="sc-44f25114-2 WJnhb"
      >
        <button
          class="sc-44f25114-3 lkgOrW"
          type="button"
        >
          Order
        </button>
        <button
          class="sc-44f25114-3 hXMIxa"
          type="button"
        >
          Return
        </button>
        <button
          class="sc-44f25114-3 lkgOrW"
          type="button"
        >
          System
        </button>
      </div>
      <form
        action="https://pc-connect.co.uk/tracking/trackreturn"
        class="sc-44f25114-4 iXxCCw"
        method="post"
      >
        <label
          for=":r3:-invoice"
          hidden=""
        >
          Invoice Number
        </label>
        <input
          autocomplete="off"
          class="sc-44f25114-5 dtSDlf"
          id=":r3:-invoice"
          maxlength="10"
          name="invoiceNumber"
          placeholder="Invoice No."
          required=""
          type="text"
        />
        <label
          for=":r3:"
          hidden=""
        >
          RMA Number
        </label>
        <input
          autocomplete="off"
          class="sc-44f25114-5 dtSDlf"
          id=":r3:"
          maxlength="10"
          name="rmaNumber"
          placeholder="RMA No."
          required=""
          type="text"
        />
        <button
          aria-label="Submit"
          class="sc-44f25114-6 mIroD"
          title="Submit"
          type="submit"
        >
          <svg
            fill="white"
            height="19"
            viewBox="0 0 384 512"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L274.7 256 105.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Delivery component renders a system build tab where build progress can be checked 1`] = `
<DocumentFragment>
  <div
    class="sc-44f25114-0 egPNsa"
  >
    <h3
      class="sc-9bf04a45-3 jXmISH"
    >
      Delivery / Tracking Information
    </h3>
    <div
      id="delivery"
      style="overflow: hidden; height: auto;"
    >
      <p
        class="sc-44f25114-1 cwwZoe"
      >
        Order now to receive by tomorrow.
      </p>
      <div
        class="sc-44f25114-2 WJnhb"
      >
        <button
          class="sc-44f25114-3 lkgOrW"
          type="button"
        >
          Order
        </button>
        <button
          class="sc-44f25114-3 lkgOrW"
          type="button"
        >
          Return
        </button>
        <button
          class="sc-44f25114-3 hXMIxa"
          type="button"
        >
          System
        </button>
      </div>
      <form
        action="https://pc-connect.co.uk/tracking/trackbuild"
        class="sc-44f25114-4 iXxCCw"
        method="post"
      >
        <label
          for=":r6:-invoice"
          hidden=""
        >
          Invoice Number
        </label>
        <input
          autocomplete="off"
          class="sc-44f25114-5 dtSDlf"
          id=":r6:-invoice"
          maxlength="10"
          name="invoiceNumber"
          placeholder="Invoice No."
          required=""
          type="text"
        />
        <label
          for=":r6:"
          hidden=""
        >
          Build Number
        </label>
        <input
          autocomplete="off"
          class="sc-44f25114-5 dtSDlf"
          id=":r6:"
          maxlength="10"
          name="buildNumber"
          placeholder="Build No."
          required=""
          type="text"
        />
        <button
          aria-label="Submit"
          class="sc-44f25114-6 mIroD"
          title="Submit"
          type="submit"
        >
          <svg
            fill="white"
            height="19"
            viewBox="0 0 384 512"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L274.7 256 105.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Delivery component renders an order tab where order tracking can be checked 1`] = `
<DocumentFragment>
  <div
    class="sc-44f25114-0 egPNsa"
  >
    <h3
      class="sc-9bf04a45-3 jXmISH"
    >
      Delivery / Tracking Information
    </h3>
    <div
      id="delivery"
      style="overflow: hidden; height: auto;"
    >
      <p
        class="sc-44f25114-1 cwwZoe"
      >
        Order now to receive by tomorrow.
      </p>
      <div
        class="sc-44f25114-2 WJnhb"
      >
        <button
          class="sc-44f25114-3 hXMIxa"
          type="button"
        >
          Order
        </button>
        <button
          class="sc-44f25114-3 lkgOrW"
          type="button"
        >
          Return
        </button>
        <button
          class="sc-44f25114-3 lkgOrW"
          type="button"
        >
          System
        </button>
      </div>
      <form
        action="https://pc-connect.co.uk/tracking/trackorder"
        class="sc-44f25114-4 iXxCCw"
        method="post"
      >
        <label
          for=":r0:-invoice"
          hidden=""
        >
          Invoice Number
        </label>
        <input
          autocomplete="off"
          class="sc-44f25114-5 dtSDlf"
          id=":r0:-invoice"
          maxlength="10"
          name="invoiceNumber"
          placeholder="Invoice No."
          required=""
          type="text"
        />
        <label
          for=":r0:"
          hidden=""
        >
          Postcode
        </label>
        <input
          autocomplete="off"
          class="sc-44f25114-5 dtSDlf"
          id=":r0:"
          maxlength="10"
          name="postcode"
          placeholder="Postcode"
          required=""
          type="text"
        />
        <button
          aria-label="Submit"
          class="sc-44f25114-6 mIroD"
          title="Submit"
          type="submit"
        >
          <svg
            fill="white"
            height="19"
            viewBox="0 0 384 512"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L274.7 256 105.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
